valhalla/filesystem.h

```diff
@@ -416,22 +422,53 @@ inline void resize_file(const path& p, std::uintmax_t new_size) {
 }
 
 inline bool remove(const path& p) {
-  return ::remove(p.c_str()) == 0;
+  bool ret = ::remove(p.c_str()) == 0;
+
+  // it can occur that another thread is removing the same file
+  // object at the same time. Occasionally it will occur that
+  // while one thread is deleting (thread 1), the other thread
+  // (thread 2) will receive -1 from ::remove() and (errno==ENOENT
+  // or errno==EINVAL). And yet, the file still exists until thread 1
+  // is finished deleting. In this exact moment, if thread 2 calls
+  // stat() on the file object in question it will get 0 which means
+  // "file object exists". We kinda want thread 2 to wait for the
+  // file object to truly go out of existence. Hence, this spin loop.
+  if (ret == false) {
+    if ((errno == EINVAL) || (errno == ENOENT)) {
+      const int max_tries = 10000;
+      int tries = 0;
+      struct stat s;
+      while ((stat(p.c_str(), &s) == 0) && (tries < max_tries))
+        tries++;
+    } else
+      throw std::runtime_error("filesystem_error filesystem::remove");
+  }
+
+  return ret;
 }
 
-inline bool remove_all(const path& p) {
+inline std::uintmax_t remove_all(const path& p) {
```

Comment:
Does this need to be inline?

Reply:
yes because its a free function in the header